torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -696,7 +696,10 @@ void IndexCompute::run() {
 }
 
 kir::Val* IndexCompute::getExtent(kir::IterDomain* id) {
-  if (extent_map_.find(id) != extent_map_.end()) {
+  if (GpuLower::current()->getPaddedParallelTypeSet().count(
+          id->parallelType())) {
+    return kir::NamedScalar::getParallelDim(id->parallelType());
```

Comment:
I wonder if this is really always correct. Since `blockDim.x` is the maximum of all domains parallelized with `TIDx`, it could be larger than a padded extent. Using `blockDim.x` in such a case doesn't seem correct.

However, I suspect it won't matter since the extent shouldn't be used in indexing. Parallelized axes of non-global tensors are marked as "zero" and are not used in indexing expressions.

Reply:
I see. For parallel access on local mem should be ok. 

I will add a few more tests on shared mem, which we would likely want. 

I guess this case has excluded padded serial `Iterdomains`, which would require over-subscription on local buffer allocation. This doesn't look likely on warp reduce related use of padding but we could address in a follow up.

Reply:
I'm not sure if we would need to care shared memory for this context. Warp shuffle is to do avoid smem-access latencies and synchronization overheads, so why would you want to store the input to the reduction in shared memory? That seems highly unlikely.

Dealing with just the Local memory would be probably just fine and make things a little simpler. For example, for supporting shared memory, the allocation logic would probably need some change as well as this extent logic in indexing.

If we only care Local, I don't think we need this change as it doesn't affect anything.

Reply:
Sure it makes sense not to handle sharedmem for now but this change was originally for indexing global buffers with padded parallel dim. I will add comments but meanwhile suggestions on putting it somewhere else also appreciated.

Reply:
Do we pad gmem tensors? I thought no.

Reply:
No this doesn't pad the tensors, this is on the indexing side. 

When we index a gmem tensor with padded dim we wanted to prefer filling the padded dimension, eg:

```cpp
T1_l = SET(T0_g[I0])

I1, I2 = split(I0,8);
I2->parallelize(TIDx)
I2->padTo(32);
```

should generate index pattern like:
```
For ...
  If ...
      T1_l[...] = T0_g[i1*blockDim.x + i2]
```

Instead of 

```
For ...
  If ...
      T1_l[...] = T0_g[i1*8 + i2]
```

Addressing this here was based on the understanding that:

1. The padded IterDomains should all be leaf domains since they have parallel types
2. The extent here is used on the indexing generation and not the bounds used by predicate

I'd really appreciate any suggestions on making this more robust or more widely applicable.

Reply:
Hmm, actually, when `blockDim.x` becomes larger than the padded extent, I suspect `warpReduceTIDX` wouldn't work unless `blockDim.x` is still a multiple of 32. The shuffle-based per-warp reduction might attempt to access non-existing warp lanes.

When a fusion has `padTo(32)`, I think it's reasonable to error out if `blockDim.x` turned out to be larger than 32. When `padTo(-1)`, we could take the maximum of all `TIDx`-mapped extents and then round up to a multiple of 32, though it's a little overwhelming. Maybe, it would just fine to disallow any other mapping that's larger than the minimal multiple of 32.

https://github.com/csarofeen/pytorch/pull/976/files#diff-6f1e7ce30c1600288756ba97ab7b4324993d2959452d874f0d15e79a16e1f850R432

Reply:
Yes I agree that tensor with larger size on the already padded dimension can be problem. I updated the binding scheme to bind block dim to largest iterdomain padded to multiple of 32 and max out at 1024. This could definitely be improved more with a systematic approach on analyzing the extents.

Reply:
OK. Back to your example above (`T1_l = SET(T0_g[I0])`), is this really supported? In `getMaybeWarpReductionDim`, there's a check that both the input and output of a reduction are on Local. That would imply the reduction input must also be parallelized with TIDx on the matching axis of the padded axis. It seems we should never hit this condition for indexing.

I'm not sure what this would mean for predicate generation. Looks like all tests result in just `true` predication. Let me dig into it a little more.